### PR TITLE
Fix seven widget defects, including the cookies the separate-domain setup needs

### PR DESCRIPTION
**Pull request by paskal**
Supersedes #2195, which carried the path-filter commit and a note describing the double pin that this PR removes outright.

## One manifest for the frontend

`frontend/` held a `package.json`, a `pnpm-workspace.yaml` and the lockfile for a workspace of exactly one package. Two manifests meant two places to declare a version, and `preact` and `@babel/core` were each written in both.

The app pin was not the one that decided. Measured with `pnpm install --lockfile-only` on a copy of the workspace, before making the change:

| app manifest | override | what installs | what the lockfile records as the app's specifier |
|---|---|---|---|
| `10.28.0` | `10.29.8` | `10.29.8` | `10.29.8` |
| `10.29.8` | `10.28.0` | `10.28.0` | `10.28.0` |

So bumping the app manifest alone was a silent no-op that left it misstating what was installed, and `pnpm install --frozen-lockfile` still passed with the two naming different versions.

Everything pnpm reads now lives in `frontend/apps/remark42`: dependencies, `packageManager`, `engines` and the 68 `pnpm.overrides` entries. `frontend/` keeps `.nvmrc`, `.husky` and `CLAUDE.md`, none of which pnpm reads.

The directory nesting stays, because every path in the repository points at `frontend/apps/remark42`. Moving the package up to `frontend/` would have rewritten 14 files and 52 lines, including `.github/dependabot.yml` and the published contributing docs, for the same result.

**The move and the refresh are separate commits.** `Drop the frontend workspace root` seeds the lockfile from the existing one and reconciles its importers, so the resolved set is identical there: 1446 packages before and after, none added, none removed. `Refresh the frontend lockfile` then re-resolves on top.

That refresh is not a version bump. **No direct dependency changes version**; the five that appear to differ change only their peer suffix. The old resolutions carried optional peer subtrees the tree no longer reaches, `ts-node` under jest, `@swc/core` under webpack, `vitest` under `@testing-library/jest-dom` and `tslib` under `webpack-dev-server`, none of them referenced by any config or source file here, since jest reads `jest.config.mjs` and webpack uses babel. 137 packages leave and 59 move within ranges the manifest already permits, 1446 down to 1308.

The build output is unaffected. Every file `pnpm build` produces is identical in size before and after. Four files differ in content, and two builds from one unchanged lockfile and `node_modules` produce the same four, so that is the css-module token churn `frontend/CLAUDE.md` documents rather than anything the refresh did.

Verified: `pnpm install --frozen-lockfile`, lint, type-check, translation-check, 416 tests and a build, all from the app directory; a full `docker build` with the frontend stages exercised; and the release round trip, `prepare-release-assets.sh` producing 53 embedded assets with the `{% REMARK_URL %}` substitution applied, then `cleanup-release-assets.sh` restoring the tree.

The published contributing docs told readers to run `pnpm i`, `pnpm build` and `pnpm dev:app` inside `./frontend`. All three would have broken, and `dev:app` was a workspace-root script that no longer exists, so they now name the app directory and `pnpm dev`.

## Markdown-only changes were starting the e2e, release and site workflows

`ci-backend.yml`, `ci-build.yml` and `ci-frontend.yml` all end their path filters with `!**.md`. The e2e workflow did not, so any markdown under `frontend/` or `backend/` matched its `frontend/**` and `backend/**` entries and started a docker build and the full browser suite.

`release.yml` had the same hole plus two of its own. It names `README.md` and `LICENSE` deliberately, since `.goreleaser.yml` packages both into the archives, so a blanket `!**.md` would have excluded them; it now excludes markdown under `backend/` and `frontend/` only. `CLAUDE.md` and the getting-started installation page were listed as well, both from the workflow's first commit in #2070. Neither is packaged, and outside their own trees the only reference to either in the repository was that filter, so the job they started could not check anything about them.

`ci-site.yml` goes on matching markdown, which is right, since the site is built from it. It excludes two names: `CLAUDE.md`, so a future `site/CLAUDE.md` cannot start a site build, and `site/README.md`, which documents how to build the site rather than being part of it.

Checked by matching every workflow's path filter against all 133 tracked markdown files, plus hypothetical `CLAUDE.md` paths under `site/`, `backend/` and `e2e/`.

## A check for the versions the docs state

The installation page tells a reader a source build needs Go 1.25, Node 24+ and PNPM 10. Nothing kept those in step with `backend/go.mod`, `engines.node`, `packageManager` and `.nvmrc`, and the drift is silent, since a wrong version in the docs builds and tests exactly as well as a right one. `.nvmrc` is the pin with form here: it sat at 16 through the whole node 20 migration because nothing red ever pointed at it.

`scripts/check-documented-versions.sh` compares each stated version against its source, and holds `.nvmrc` to `engines.node`. It fails when the page states no version at all, so deleting the claims cannot turn it into a check that passes by comparing nothing.

Its own workflow rather than a step in an existing one, since the inputs span the backend module, the frontend manifest and the site.

Each of the five failure modes was exercised against the real tree: Go, Node and PNPM stated wrongly, `.nvmrc` drifting from `engines.node`, and the claims removed entirely. All five fail, and the unmodified tree passes.

## Seven defects found while the frontend was open, fixed here rather than in PRs of their own

Two share an origin: `5825a55b`, the January 2021 frontend rewrite, first released in v1.7.0.

**#2201, the cookie fallback led nowhere.** When third-party cookies are blocked the auth panel offers "open comments in new page", pointing at `/web/comments.html`. That rewrite removed the build entry for the page while leaving both `templates/comments.ejs` and the link in place, so the one reader who cannot sign in any other way has been getting a 404 for four years. Confirmed against a running v1.16.4 binary and a build of master, both 404. The template needed no changes. An end-to-end case now opens the page on a thread carrying a comment and waits for that comment, so the page being served, its inline script running and it asking for the thread named in its own query string are all covered; against an image built without the plugin entry it fails on the 404.

**#2203, path-prefixed installations fetched their icons from the wrong place.** The same rewrite fixed the public path to the domain root, so an instance at `https://example.com/remark42`, which `manuals/separate-domain` documents, requested `https://example.com/web/google.svg`. Fifteen provider icons in `remark.mjs` and one in `last-comments.mjs`. The path now derives from the URL the bundle was loaded from, correct for both arrangements, and the file loader no longer overrides it. Verified in the built bundle: no root-relative asset string survives.

**#2202, the host page accepted messages from any frame.** Every frame on a page can reach `window.parent`, and the handler in `embed.ts` resizes the widget, scrolls the page and opens the profile overlay without checking who sent the message. An advertisement in a sibling iframe could drive all three. It now ignores anything that did not come from a frame this module created. Scope, to be exact about it: `createIframe` always builds `src` from `BASE_URL`, so no forged message could ever point the overlay at another origin, and no comment or token data was reachable.

**The OAuth flow never tore its polling down.** `subscribed` is declared, checked and cleared in `auth.api.ts`, but nothing ever sets it, so the guard against a second subscription is dead code and every provider click attaches another `visibilitychange` and `focus` pair. The five minute deadline then rejects without unsubscribing, and the one minute retry reschedules itself for as long as `getUser` returns null. Cross-domain embedding is exactly where `getUser` never stops returning null, so a reader on the arrangement `manuals/separate-domain` documents is left polling `/auth/user` once a minute for the life of the page, against a route capped at 2 req/s. The rejection also carried no argument, and `auth.tsx` stores whatever it catches as the error state, so the interface had `undefined` to render. The deadline now tears the subscription down and rejects with an error. Three tests cover it, and all three fail with the teardown removed.

**The widget document trusted any sender too.** Hardening the parent only covered one direction. `root.tsx` acts on `signout` and `theme`, and `remark.tsx` on `theme`, straight out of `parseMessage`, so anything holding a reference to the frame could sign a reader out. `auth.hooks.ts` already checked `event.source !== window.parent`; that check is now a shared `isFromParent` and the three listeners that lacked it use it as well. Origin cannot stand in for it: the host page is whatever site embeds the widget, and `ALLOWED_HOSTS` is enforced server side through `frame-ancestors`.

**`createInstance` stacked its listeners.** It reuses the marked iframe rather than building one, but installed three listeners plus a title observer on every call, while `destroy` could only reach the newest closure. A second call without a `destroy` therefore stranded a set permanently, and the reader got two reactions to every resize. The current instance's listeners are now detached before the next set goes on. Reuse and the ignored config are deliberately unchanged: that contract is an open question in `docs/backlog/public-widget-contract-docs-and-types.md`, not something to settle in a bugfix.

**The auth cookies could not be delivered in the arrangement they exist for.** remark42 on its own name serving a site on another one is documented in `manuals/separate-domain`, and the client's cookie writer broke it twice. `setAuthCookie` prefixed with `__Host-` whenever the page was https, so a real deployment wrote `__Host-JWT` and `__Host-XSRF-TOKEN`, while the backend looks for `JWT` and the fetcher reads `XSRF-TOKEN`; nothing anywhere reads a prefixed name. Nothing caught it because the prefix comes from the page protocol and every test and the dev server run on http, so there is now a second suite pinned to an https page. Both cookies were also `SameSite=Strict`, and `SameSite` is judged against the top-level site rather than the request's own origin, so a Strict cookie is never sent from a third-party frame. They now follow the embedding: Strict while the widget shares its page origin, `None` with `Secure` and `Partitioned` once it does not. That leaves the client half of #1877 working, whose reporter wanted `AUTH_SEND_JWT_HEADER` for exactly this arrangement; the server's own cookies still carry no `Partitioned`, which is upstream work in `go-pkgz/auth`.

**The injection review found, reachable only because of the first defect.** `templates/comments.ejs` built its title with `innerHTML` from the `url` query parameter, so restoring the build entry made a reflected XSS live on the instance origin: the page is a top-level document, `frame-ancestors` does not apply to it, and the `/web` CSP allows `unsafe-inline`. The anchor is built through the DOM with `textContent` now, and only `http(s)` reaches `href`, since escaping alone leaves a `javascript:` url working. Two e2e subtests pin both halves, mutation-checked separately: restoring `innerHTML` fails four assertions, while keeping the escaping and dropping only the scheme guard fails the `href` one alone.

Review also caught that the poll teardown test never dispatched the events that reach the poll, so no request fired and it compared zero to zero, passing with the fix reverted. It drives the poll now. And the teardown could not cancel an in-flight `getUser`; a closure-local flag after the await does, chosen over a second guard at the top of the handler because only one of the two is detectable by mutation.

436 frontend tests, lint, type-check and a production build all pass, every size budget is met and `data-testid` is absent from every bundle. Each fix has tests that fail when it is reverted, checked one mutation at a time rather than in aggregate.

## Two documents

`docs/backlog/public-widget-contract-docs-and-types.md` and `docs/plans/2026-08-19-frontend-simplification.md` were untracked working files. Both are in, with their `site/src/docs` paths corrected to the hugo layout.

The plan is rewritten against master rather than committed as it stood, because it aged badly in three days: three of its four incremental cards landed in #2175, #2176 and #2178, its task to extend the e2e suite is complete, and most of its measurements and file references had moved. It also predicted devDependencies falling from 82 to roughly 20 after the toolchain collapse; the figure is 59 on master and 61 once this PR folds the two root entries into the app manifest, and what survives is not compilation but eslint, stylelint, prettier, size-limit, formatjs and the webpack plugin set, so the remaining question is what verifies the widget rather than what builds it.

R1, the separate-domain arrangement, is promoted from a constraint bullet to a named requirement with acceptance criteria, since a test that signs in and posts without reloading passes while persistence is entirely broken. It also gains a section naming what neither direction has costed: a definition of done for removing npm, the browser floor once the transpiler goes, a runtime dependency inventory, a security section, translation tooling after node, what replaces the 426 unit cases, size budgets and the dev loop, storage partitioning, and the sequencing against #2128, which is rewriting the same CSS layer on the assumption that hashed class names stay.

Two conventions in it are deliberate. It is written as though #2196, #2197 and #2198 have landed, and says so in its header, because costing either direction against the state before them would be costing a world that no longer exists. And it cites code by symbol rather than by line: ninety-seven `file:NN` references became function, type, config-key and class names, after three of them went stale from fixes made in this same branch.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -3,10 +3,9 @@ FROM --platform=$BUILDPLATFORM node:24-alpine AS frontend-deps
 ARG SKIP_FRONTEND_TEST
 ARG SKIP_FRONTEND_BUILD
 
-WORKDIR /srv/frontend/
+WORKDIR /srv/frontend/apps/remark42/
```
[the remark this answers is not in the document]
Comment (paskal) [reply Dockerfile:9]: Fixed: the frontend-deps stage now sets `ENV CI=true`, matching what the build stage below it already does, so `prepare` skips instead of trying to install husky hooks into a build context with no git repository.

Worth noting the reach: the deps stage installed `git` for pnpm, and the image built green both before and after, so this was not failing. It was doing pointless work in a layer that exists to be cached.

File: site/content/docs/contributing/frontend/index.md [hunk 2]
```diff
@@ -23,7 +23,7 @@ You must have at least 2GB RAM or swap enabled for building.
 
 - install [Node.js 24](https://nodejs.org/en/) or higher (we recommend using [NVM](https://github.com/nvm-sh/nvm) for node version autoswitch)
 - install [PNPM 10](https://pnpm.io/installation)
-- run `pnpm i` inside `./frontend`
+- run `pnpm i` inside `./frontend/apps/remark42`
 
 Running `pnpm i` will set up pre-commit hooks into your git repository. They are used to reformat your frontend code using `prettier` and lint with `eslint` and `stylelint` before every commit.
```
[the remark this answers is not in the document]
Comment (paskal) [reply site/content/docs/contributing/frontend/index.md:28]: Half fixed, half deliberate.

`CLAUDE.md` was right and is fixed. Three lines told a reader to run `cd frontend && pnpm dev:app`, `pnpm test` and `pnpm lint`, all of which break here, and `dev:app` no longer exists at all since it was a workspace-root filter script. They now name `frontend/apps/remark42` and `pnpm dev`.

`docs/plans/completed/2026-02-24-bem-to-css-modules-batch1.md` I have left alone. It is an archived record of a plan that was executed in February, and its `cd frontend && pnpm lint` lines are ticked-off checklist items describing commands that were correct when they were run. Rewriting them would make the record say something that did not happen, and it is not guidance anyone follows now.

Swept every tracked markdown file for the pattern rather than fixing only the two named: nothing else points pnpm at the old root.

**Comment by umputun** [comment]
one blocker, three non-blocking.

**restoring the comments.html entry makes a reflected XSS reachable.** `templates/comments.ejs:66` builds the title with `innerHTML`:

```js
titleElement.innerHTML = 'Comments for <a href="' + query.url + '">' + query.url + '</a>';
```

`query.url` comes straight out of `decodeURIComponent` on the query string, and the /web CSP (`backend/app/rest/api/middleware.go:273`) is `script-src 'self' 'unsafe-inline'`, so `?site_id=remark&url=%22%3E%3Cimg%20src%3Dx%20onerror%3Dalert(document.domain)%3E` executes on the instance origin. `frame-ancestors` does not apply, it is a top-level navigation.

the hole is as old as the template, but the page 404s on master, so this PR is what makes it reachable. Building the anchor with DOM APIs and setting the text through `textContent` closes the injection; the `href` needs a scheme check too, or a `javascript:` URL survives the fix.

**the poll teardown test does not exercise the poll.** `auth.api.spec.ts` "stops polling once it gives up" mocks `window.open` and `document.hasFocus`, but never dispatches `focus` or `visibilitychange`. `handleWindowVisibilityChange` is reachable only from those two listeners and from the retry timer it sets itself, so no fetch ever fires: `afterGivingUp` is 0, and the final assertion compares 0 to 0. It passes with the teardown reverted. "detaches its listeners" is the one carrying the weight.

**the teardown cannot cancel an in-flight `getUser`.** `auth.api.ts:61` awaits, and a null resolving after `giveUp` has run reaches lines 65-69 and schedules a fresh 60s retry with nothing left to clear it, so the poll this fixes can still run for the life of the page. A closure-local `stopped` flag checked after the await closes it. The global `subscribed` will not, since a second `oauthSignin` can start while the old request is still resolving.

**one message listener still trusts any sender.** the inline handler in `templates/iframe.ejs:76-91` acts on `data.hash` through `location.replace` and on `data.title`, with no source check. A sibling frame can post a comment hash, `Root.checkUrlHash` reads it as local state and emits `scrollTo`, and the newly hardened parent accepts that because it came from its own iframe. Impact is a forced scroll or a changed child title, so not a blocker, but it is the same class as #2202 and `event.source === window.parent` closes it.

on ordering: #2196 goes in first, so this one needs a conflict resolution in `e2e/widgets_test.go`. `signInAnon` gains a page argument there, and `TestWidgets_CommentsPageOpensAThreadOnItsOwnOrigin` calls the 3-arg form. `go vet -tags=e2e` catches it if the resolution keeps the stale call.

**Comment by paskal** [comment]
All four addressed, and the second one was the important catch.

**The injection.** The anchor is built through the DOM now, `textContent` for the label, and only `http(s)` reaches `href` so a `javascript:` url does not survive the escaping. Verified against the page as served, with your payload: no dialog, no injected element, the markup rendered as text, and no `href` attribute at all. The `javascript:` variant gets no `href` either.

**The poll teardown test was vacuous and you are right that it passed with the fix reverted.** Zero fetches ever fired: I measured it, `afterGivingUp` was 0 and the assertion compared 0 to 0. My mutation check had caught three failures and I read the aggregate instead of checking which test carried them, which is exactly the mistake the test itself was supposed to prevent. It now dispatches `focus`, asserts requests are actually being made, asserts they keep coming across a retry interval, and only then that they stop. Reverting the teardown fails it.

**The in-flight `getUser`.** Added, as a closure-local flag for the reason you gave. Worth reporting what the mutation testing showed, because it changed the shape: I first guarded both at the top of the handler and after the await, and neither mutation alone could be detected, since each covered the other. Guarding only after the await is the better single fix, because it stops the stray timer being created at all rather than neutering it later, so that is what is in. Removing it now fails `does not reschedule when getUser resolves after the deadline`.

**The iframe template handler.** `event.source !== window.parent` and return, matching the host page side.

**Ordering.** Rebased onto master with #2196 in. The conflict was where you said: `signInAnon` gained its page argument and `TestWidgets_CommentsPageOpensAThreadOnItsOwnOrigin` called the 3-arg form. Resolved and `go vet -tags=e2e` is clean, which is what would have caught it.

436 frontend tests, lint, type-check and a production build all pass, `comments.html` is emitted and `data-testid` is absent from every bundle.

**Comment by paskal** [comment]
Rebased on master with #2196 in, and squashed to three commits: the workspace-root change, the CI change, and all the defects with the e2e coverage as one. Tree byte-identical to before the squash, verified by diffing it.

All four review points are addressed, detail in my earlier comment. Short version: the injection is closed by building the anchor through the DOM with a scheme check and pinned by two mutation-checked e2e subtests, the poll teardown test now actually drives the poll instead of comparing zero to zero, the in-flight `getUser` is guarded, and the iframe template handler takes only the parent.

The conflict you predicted was exactly where you said, `signInAnon` gaining its page argument, resolved with `go vet -tags=e2e` clean.

**Review by umputun — APPROVED** [review]
